File: python/endpoints/docking/update_dock.py
from clearpath_config.common.utils.yaml import read_yaml
import rclpy
from rclpy.node import Node
from clearpath_dock_msgs.srv import UpdateDock

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'UpdateDock'


class UpdateDock(Node):

    def __init__(self, namespace: str):
        super().__init__('update_dock_client')
        self._update_dock_client = self.create_client(UpdateDock, f'/{namespace}/docking/dock_manager/update')

        while not self._update_dock_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/docking/dock_manager/update service not available, waiting again...')

    def update_dock(self):
        req = UpdateDock.Request()
        self.future = self._update_dock_client.call_async(req)
        self.future.add_done_callback(self.update_dock_srv_response_cb)

    def update_dock_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Dock updated successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to update dock: {response.message}')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    update_dock_client = UpdateDock(namespace)
    update_dock_client.update_dock()

    rclpy.spin(update_dock_client)


if __name__ == '__main__':
    main()

File: cpp/endpoints/docking/update_dock.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_dock_msgs/srv/update_dock.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("update_dock_client");
  rclcpp::Client<clearpath_dock_msgs::srv::UpdateDock>::SharedPtr client =
    node->create_client<clearpath_dock_msgs::srv::UpdateDock>("/" + namespace_ + "/" + std::string("docking/dock_manager/update"));

  auto request = std::make_shared<clearpath_dock_msgs::srv::UpdateDock::Request>();

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("update_dock_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("update_dock_client"), "/%s/docking/dock_manager/update service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("update_dock_client"), "Dock updated successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("update_dock_client"), "Failed to update dock!");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("update_dock_client"), "Failed to call service /%s/docking/dock_manager/update", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}
